### DISCUSSION TITLE
【新提案】引入一个模块时，可以排除某些属性，采用  import   except， 方便做Tree Shaking

### DISCUSSION THREAD

**Discussion by wenber** [JSCIG/es-discuss#27] [Ideas]
有这样一种场景，模块A导出了5个功能函数， 我在模块B中依赖了模块A中的4个函数。
当前规范下有两种写法

```javascript
// 虽然我只调用4个，但会全量引入，不推荐这样写
import  *   as   A   from   './A';    
// 推荐这样写，但比较麻烦，如果A包含10个函数，B依赖了9个，简直太啰嗦了
import {A1, A2, A3, A4} from './A';  
```
期望能在引入时，通过规范反向排除某些模块，比如这里，如果我不需要A5,那么可以这样写

```javascript
 import  *   expect  {A5}  as  A   from './A';
```
通过规范让编译阶段可以排除A5,做到按需引入。
欢迎大家讨论。

**Comment by wenber** [comment]
更正下，上面有个单词写错了：   import  *   except  {A5}  as  A   from './A';   提案内容区居然发布了就没法编辑了

**Reply by wenber** [reply]
@hax 贺老师，求看看哈

**Comment by wenber** [comment]
请大家把脉下，ecma发展历程里有很多都有正反相对的用法，比如indexOf vs lastIndexOf、 startswith  vs  endswith

**Comment by LongTengDao** [comment]
用于 `export * except {} from './A';` 可能会比较有用。`import` 本身是能静态分析的，本身就支持 tree-shake。

**Reply by wenber** [reply]
import是支持，但如果写成  import  *   as   A   from   './A';    只调用了A.a()； A里的其他方法会被引入不？

**Reply by LongTengDao** [reply]
目前主流的构建工具（rollup）会跟踪`a`的副作用引用，无关的都不会被引入，你可以试一下rollup。事实上，你的需求正是es6 import/export引入的主要目的和优势。

**Comment by hax** [comment]
如果希望明确使用了哪些导入，那么通过 `import {name1, name2, ...} from mod`来穷举是最精确的。注意js并不支持`import * from mod`，其原因是这种语法要求必须先加载mod完成之后才能确定名字解析，对于web来说是行不通的（严重降低解析和加载的并行性）。故而js只支持`import * as ns from mod`。

那么对于`import * as ns from mod`来说，`ns`已经是一个对象了，再标明ns里有什么或没有什么，价值就不高了。

如果我们的目的是静态编译优化，那么由优化器自行静态分析来判断到底使用了`ns`下哪些名字，可能才是最直接的方式。

如果一定想要明确，那么通过工具函数进行对象操作，比如 `const ns1 = except(ns, ['A', 'B'])` ，配合静态分析，也可以做到。

我也赞同 @LongTengDao 的观点，如果一定要说的话，`export * except {A} from mod` 大概是更有用的。如果真的有这个之后，那么也许可以加上 `import * except {A} from mod` 作为对称。

**Reply by LongTengDao** [reply]
这么顺着想下来的话，这种情况下再引入`import * except { A } as ns from mod`或许有一个用，就是需要把`ns`对象作为一个整体传递给某个API（`use(ns)`），这种情况下，虽然只用到其中一部分，但剩下的却无法tree shake了（且不论`const { A: neverUse, ...ns$1 } = ns;`是否只能到被tree shake，要跟踪属性更改更是无法walkaround……）。

不过这不是题主的原始语境，那种无`as`的`*`引入太可怕了……感觉随时会把自己炸飞

**Comment by troy351** [comment]
事实上 webpack 很早就支持了对于 `import * as A from './a'` 的 treeshake
https://github.com/webpack/webpack/issues/2713
